File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/entity/component/EntityComponentActorFactory.java
package net.bestia.zoneserver.actor.entity.component;

import akka.actor.AbstractActor;
import akka.actor.ActorContext;
import akka.actor.ActorRef;
import net.bestia.entity.EntityService;
import net.bestia.entity.component.ActorSync;
import net.bestia.entity.component.Component;
import net.bestia.zoneserver.actor.SpringExtension;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Objects;

/**
 * Depending on the given component ID this factory will create an actor
 * suitable for the component. This is done by checking the annotation
 * of the component.
 * 
 * @author Thomas Felix
 *
 */
@org.springframework.stereotype.Component
public class EntityComponentActorFactory {

	private static final Logger LOG = LoggerFactory.getLogger(EntityComponentActorFactory.class);

	private final EntityService entityService;

	@Autowired
	public EntityComponentActorFactory(EntityService entityService) {

		this.entityService = Objects.requireNonNull(entityService);
	}

	/**
	 * Starts a component actor which is responsible for managing continues
	 * callback to some component code.
	 * 
	 * @param componentId
	 *            The ID of the already saved and existing component to create
	 *            an actor for.
	 * @return The created actor or null if something went wrong.
	 */
	public ActorRef startActor(ActorContext ctx, long componentId) {

		final Component comp = entityService.getComponent(componentId);

		if (comp == null) {
			LOG.warn("Component {} does not exist. Can not build component actor for it.", componentId);
			return null;
		}

		return startActorByAnnotation(ctx, comp);
	}

	private ActorRef startActorByAnnotation(ActorContext ctx, Component comp) {
		if (!comp.getClass().isAnnotationPresent(ActorSync.class)) {
			LOG.warn("Component {} (id: {}) has no ComponentActor annotation. Can not create Actor.",
					comp.getClass().getName(), comp.getId());
			return null;
		}

		final ActorSync compActor = comp.getClass().getAnnotation(ActorSync.class);
		try {
			@SuppressWarnings("unchecked")
			final Class<? extends AbstractActor> actorClass = (Class<? extends AbstractActor>) Class.forName(compActor.value());
			
			final ActorRef actorRef = SpringExtension.actorOf(ctx,
					actorClass,
					null,
					comp.getEntityId(),
					comp.getId());

			LOG.debug("Starting ComponentActor: {} ({}) for entity: {}.",
					actorClass.getSimpleName(),
					comp.getId(),
					comp.getEntityId());

			return actorRef;
		} catch (ClassNotFoundException e) {
			LOG.warn("Could not start ComponentActor. Class not found: {}.", compActor.value());
			return null;
		}
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/entity/component/EntityComponentActorFactory.kt
package net.bestia.zoneserver.actor.entity.component

import akka.actor.AbstractActor
import akka.actor.ActorContext
import akka.actor.ActorRef
import net.bestia.zoneserver.actor.ActorComponent
import net.bestia.zoneserver.actor.SpringExtension
import net.bestia.zoneserver.entity.component.Component
import org.reflections.Reflections
import kotlin.reflect.KClass
import org.springframework.stereotype.Component as SpringComponent

/**
 * Depending on the given component class this factory will create an actor
 * suitable for the component. This is done by checking the annotation
 * of the component.
 *
 * @author Thomas Felix
 */
@SpringComponent
class EntityComponentActorFactory {

  private val componentToActorClass: Map<KClass<out Component>, Class<out AbstractActor>>

  init {
    val reflections = Reflections(javaClass.`package`.name)
    val annotated = reflections.getTypesAnnotatedWith(ActorComponent::class.java)
    componentToActorClass = annotated.asSequence()
        .filter { AbstractActor::class.java.isAssignableFrom(it) }
        .map {
          @Suppress("UNCHECKED_CAST")
          it as Class<out AbstractActor>
        }
        .map {
          val annotation = it.getAnnotation(ActorComponent::class.java)
          annotation.component to it
        }.toList()
        .toMap()
  }

  fun startActor(ctx: ActorContext, component: Component): ActorRef? {
    return startActorByAnnotation(ctx, component)
  }

  private fun startActorByAnnotation(
      ctx: ActorContext,
      component: Component
  ): ActorRef {
    val actorClass = componentToActorClass[component::class]
        ?: throw IllegalArgumentException("Could not find actor for component class: ${component.javaClass}")

    val hasCtorWithComponentAsArg = actorClass.declaredConstructors.any {
      it.parameterTypes.any { paramType -> Component::class.java.isAssignableFrom(paramType) }
    }

    return if (hasCtorWithComponentAsArg) {
      SpringExtension.actorOf(ctx, actorClass, component)
    } else {
      throw IllegalStateException("Actor $actorClass does not use component ${component.javaClass} as ctor argument.")
    }
  }
}
